### ISSUE TITLE
Feature Request: Consider adding support for the OAuth2 Device Authorization Flow

### ISSUE THREAD

**Issue by MarkJaroski** [abraunegg/onedrive#2693]
### Is your feature request related to a problem? Please describe.

OAuth2 has a device authorization flow which could provide a smoother authentication method compared to the native application flow which the project uses at the moment.

### Describe the solution you'd like

The user should be able to specify the device authorization flow as an option. In this flow the native application presents an 8 character code to the user, who then enters it online, which either authorizes the application against an existing session or authenticates the user.

### Describe alternatives you've considered

Well, you could eventually pop up a browser window, but that would be a lot harder to achieve.

**Comment by kevinchristianson** [comment]
I'm trying to auth in ubuntu server and am struggling to complete the loop with no GUI. I second the request for an alternate device authorization flow.

**Comment by abraunegg** [comment]
@kevinchristianson 
> I'm trying to auth in ubuntu server and am struggling to complete the loop with no GUI. I second the request for an alternate device authorization flow.

FYI - you *do not need* a GUI to complete the auth process.

You can copy the auth URL to *any* system, do the auth there, then copy the URI back to the prompt.

**Comment by MarkJaroski** [comment]
> FYI - you _do not need_ a GUI to complete the auth process.
> 
> You can copy the auth URL to _any_ system, do the auth there, then copy the URI back to the prompt.

I can absolutely verify this, and the _nativeclient_ approach isn't a problem for me at all. I've even borrowed this technique - which I first learned about here - in a set of scripts that I wrote to demonstrate OAuth2 flows for my colleagues (and vendors !)

But the device flow is easier for people to understand, so it would be nice to have.

**Comment by kevinchristianson** [comment]
Sorry for not responding earlier--haven't had a chance to get back to this. I have tried copying the URL to another system and copying the response URL back but I'm still hitting 400s. Probably something on my end though if other users aren't hitting it. Would love another login option though :)

**Comment by abraunegg** [comment]
@kevinchristianson 
Ensure you are pasting back the FULL response URI back to the application. Also ensure your system running 'onedrive' can access the required URL that validates the URI.

You can also double check what is going when you paste the repsonse  on by running the application with `-v -v --debug-https` at startup

**Comment by kevinchristianson** [comment]
I had some other issues with my server and ended up starting fresh on ubuntu desktop. I got things working using the web browser there without a hitch. Still not sure why it didn't work on ubuntu server, but I assume it was user error or something wrong on my end. Appreciate the help!

**Comment by phuzz** [comment]
Am I right in thinking that using the device authorization flow would mean not having to copy the URI back into the client? If so this would be really useful for me. I have a lot of...lets say 'untechnical'...users who struggle with the current re-authorisation process, and streamlining it in any way would be very helpful.

**Comment by abraunegg** [comment]
@MarkJaroski , @kevinchristianson , @phuzz 

Do any of your environments use / leverage Intune ?

#3209 will implement native Intune support.

Unfortunatly, to support oauth2 via device flow means the the existing azure application id this client uses, needs to be updated - however Microsoft has blocked changes to older application id's ... 

If you use Intune, when this is finalised, when you use a new application id, you will have authenication via that method.

Doea this work for you ?

**Comment by MarkJaroski** [comment]
Hi, I'm afraid I don't use Intune. Rather I was thinking of something a bit simpler. I'll put together a demo in bash to show you how the handshake works, but from the user point of view the handshake works like this:

```
To connect to onedrive please visit https://login.microsoft.com/somepath and enter the following code:

      XXX-XXXX

```

Then the client tries once every couple of seconds, and when the authorization has been granted does whatever setup is required.

I don't think Intune or any other kind of third-party agent is required.

**Comment by abraunegg** [comment]
@MarkJaroski 
Thanks for your feedback.

I will bump this to v2.5.6 release and look at this post the Intune authentication PR.

**Comment by MarkJaroski** [comment]
OK, great! It isn't a priority for me because cutting and pasting the URL isn't a problem. But less OAuth2 literate users could benefit. I support the addition with the goal of helping them.

**Comment by abraunegg** [comment]
@MarkJaroski, @kevinchristianson, @phuzz

Please can you test the PR below that implements your feature request.

First install all the require platform dependencies to build the client on your respective platforms. Please read https://github.com/abraunegg/onedrive/blob/master/docs/install.md#building-from-source---high-level-requirements and then follow correctly for your platform.

> [!IMPORTANT]
> Currently there is an issue with installing the DMD or LDC compiler using the documentation instructions:
> `curl -fsS https://dlang.org/install.sh | bash -s dmd` or `curl -fsS https://dlang.org/install.sh | bash -s ldc`
>
> This is not something I can solve and I have emailed the appropriate folk and raised https://github.com/dlang/dmd/issues/20830
> 
> Please augment the installation of DMD or LDC with your distribution version if it is available

Once this is done, to clone the PR to resolve your issue, you can use a script like the following:
```
#!/bin/bash

PR=3313

rm -rf ./onedrive-pr${PR}
git clone https://github.com/abraunegg/onedrive.git onedrive-pr${PR}
cd onedrive-pr${PR}
git fetch origin pull/${PR}/head:pr${PR}
git checkout pr${PR}

# Configure and Build
./configure --enable-debug --enable-notifications; make clean; make;
./onedrive --version
```

This script will create a local folder called `onedrive-pr3313` with the PR version.

To run the PR, you need to run the client from the PR build directory:
```
./onedrive <any other options needed>
```
To install the PR, you will need to perform `sudo make install` to install the PR version to your system.

When running the PR, your version should be: `onedrive v2.5.5-76-gc2c223c` or greater.

To enable this new capability, add the following to your config file:
```
use_device_auth = "true"
```


This PR is yet to be documented, but has been tested with a OneDrive Business Account.

I have *tried* to get this to work with a OneDrive Personal account (@outlook.com or @hotmail.com email address) and no matter what I do, this fails with the following:

![Image](https://github.com/user-attachments/assets/b974ba84-9938-4214-9099-8cd766bdea3d)


When the client authenticates using this method, expect application output similar to the following:
```
........
Using IPv4 and IPv6 (if configured) for all network operations
Attempting to contact Microsoft OneDrive Login Service
Successfully reached Microsoft OneDrive Login Service
Configuring Global Azure AD Endpoints

Authorise this application by visiting:

https://microsoft.com/devicelogin

Enter the following code when prompted: ABCDEFGHI

This code expires at: 2025-Jun-02 15:27:30.2262776

[14m 59s remaining] Still pending authorisation ...
[14m 53s remaining] Still pending authorisation ...
[14m 48s remaining] Still pending authorisation ...
[14m 43s remaining] Still pending authorisation ...
[14m 37s remaining] Still pending authorisation ...
[14m 32s remaining] Still pending authorisation ...
[14m 27s remaining] Still pending authorisation ...
[14m 21s remaining] Still pending authorisation ...
[14m 16s remaining] Still pending authorisation ...
[14m 11s remaining] Still pending authorisation ...
[14m 05s remaining] Still pending authorisation ...
[13m 59s remaining] Still pending authorisation ...
Access token acquired!

Your command line input is missing either the '--sync' or '--monitor' switches. Please include one (but not both) of these switches in your command line, or refer to 'onedrive --help' for additional guidance.

It is important to note that you must include one of these two arguments in your command line for the application to perform a synchronisation with Microsoft OneDrive
```

**Comment by MarkJaroski** [comment]
It works like a charm:

```
$ ./onedrive --reauth
Reading configuration file: /home/mark/.config/onedrive/config

WARNING: Your 'skip_file' configuration is missing important default entries. Temporary and/or transient files that would normally be skipped may now be included in syncing.

Configuration file successfully loaded
Deleting the saved authentication status ... re-authentication requested
Using IPv4 and IPv6 (if configured) for all network operations
Attempting to contact Microsoft OneDrive Login Service
Successfully reached Microsoft OneDrive Login Service
Configuring Global Azure AD Endpoints

Authorise this application by visiting:

https://microsoft.com/devicelogin

Enter the following code when prompted: DAAXZGNL8

This code expires at: 2025-Jun-02 13:03:54.1208572

[14m 59s remaining] Still pending authorisation ...
[14m 54s remaining] Still pending authorisation ...
[14m 48s remaining] Still pending authorisation ...
[14m 43s remaining] Still pending authorisation ...
Access token acquired!

Your command line input is missing either the '--sync' or '--monitor' switches. Please include one (but not both) of these switches in your command line, or refer to 'onedrive --help' for additional guidance.

It is important to note that you must include one of these two arguments in your command line for the application to perform a synchronisation with Microsoft OneDrive
```

I think if I were to change anything it would be just the visual separation, keeping in mind that this is aimed at those users who aren't perhaps as well-versed in OAuth2 flows. For instance here's what my go-mobilzon-bot does:

```
$ ./go-mobilizon-bot --authorize --authconfig test/auth.json
Please visit this URL and enter the code below https://mobilisons.ch/login/device

BCFF-GEGU

......

Access token acquired!
```


But really that's just down to taste and style. At the very least this provides GUI developers with access to this flow which is great.

**Comment by MarkJaroski** [comment]
Oh, and Thank you!

**Comment by MarkJaroski** [comment]
I'll create a personal account to try to debug that. I'm not sure what the problem could be, since most of the time personal devices are more likely to use the device code flow.

**Comment by abraunegg** [comment]
@MarkJaroski 
I spent a lot of time yesterday debugging it to essentially determine that Microsoft is blocking personal accounts from device authentication.

Pleease read:

https://github.com/abraunegg/onedrive/blob/implement-fr-2693/docs/usage.md#authorise-the-application-with-your-microsoft-onedrive-account

**Comment by MarkJaroski** [comment]
"unless the application is approved by Microsoft"

Which gives them an opportunity to shake smart-tv and console makers down for some cash.

Figures.

**Comment by abraunegg** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
